homeassistant/components/bluetooth/passive_update_coordinator.py

```diff
@@ -0,0 +1,304 @@
+"""The Bluetooth integration."""
+from __future__ import annotations
+
+from collections.abc import Callable
+import dataclasses
+from datetime import datetime, timedelta
+import logging
+import time
+from typing import Any, Generic, TypeVar
+
+from home_assistant_bluetooth import BluetoothServiceInfo
+
+from homeassistant.const import ATTR_IDENTIFIERS, ATTR_NAME
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant, callback
+from homeassistant.helpers.entity import DeviceInfo, Entity, EntityDescription
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.event import async_track_time_interval
+
+from . import (
+    BluetoothCallbackMatcher,
+    BluetoothChange,
+    async_address_present,
+    async_register_callback,
+)
+from .const import DOMAIN
+
+UNAVAILABLE_SECONDS = 60 * 5
+NEVER_TIME = -UNAVAILABLE_SECONDS
+
+
+@dataclasses.dataclass(frozen=True)
+class PassiveBluetoothEntityKey:
+    """Key for a passive bluetooth entity.
+
+    Example:
+    key: temperature
+    device_id: outdoor_sensor_1
+    """
+
+    key: str
+    device_id: str | None
+
+
+_T = TypeVar("_T")
+
+
+@dataclasses.dataclass(frozen=True)
+class PassiveBluetoothDataUpdate(Generic[_T]):
+    """Generic bluetooth data."""
+
+    devices: dict[str | None, DeviceInfo] = dataclasses.field(default_factory=dict)
+    entity_descriptions: dict[
+        PassiveBluetoothEntityKey, EntityDescription
+    ] = dataclasses.field(default_factory=dict)
+    entity_data: dict[PassiveBluetoothEntityKey, _T] = dataclasses.field(
+        default_factory=dict
+    )
+
+
+_PassiveBluetoothDataUpdateCoordinatorT = TypeVar(
+    "_PassiveBluetoothDataUpdateCoordinatorT",
+    bound="PassiveBluetoothDataUpdateCoordinator[Any]",
+)
+
+
+class PassiveBluetoothDataUpdateCoordinator(Generic[_T]):
+    """Bluetooth data update coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        address: str,
+        update_method: Callable[[BluetoothServiceInfo], PassiveBluetoothDataUpdate[_T]],
+    ) -> None:
+        """Initialize the coordinator."""
+        self.hass = hass
+        self.logger = logger
+        self.name: str | None = None
+        self.address = address
+        self._listeners: list[
+            Callable[[PassiveBluetoothDataUpdate[_T] | None], None]
+        ] = []
+        self._entity_key_listeners: dict[
+            PassiveBluetoothEntityKey | None,
```

Comment:
I think we should add a comment or extend the docstring explaining what the different kinds of keys mean here, `PassiveBluetoothEntityKey` vs `None`.

Reply:
Good catch.  I don't think we need this anymore since I split the listeners into two lists to avoid the `None` case but forgot to remove it.